content/blog/magrittr-2-0/index.Rmd

```diff
@@ -0,0 +1,195 @@
+
+Similarly, warnings that were previously issued might now be suppressed by a function you're piping into. That's because the following expressions are now almost completely equivalent:
+
+```{r}
+# Piped
+warning("foo") %>% suppressWarnings()
+
+# Nested
+suppressWarnings(warning("foo"))
+```
+
+Thanks to this change, you will now be able to pipe into testthat error expectations, for instance:
+
+```{r}
+library(testthat) %>%
+  suppressMessages()
+
+{ 1 + "a" } %>%
+  expect_error("non-numeric argument")
+```
+
+
+## Backtraces
+
+The R implementation of the magrittr pipe was rather costly in terms of backtrace clutter. This made it difficult to debug errors with functions using the pipe:
+
+```{r, eval = FALSE}
+foo <- function() bar()
+bar <- function() 1 %>% identity() %>% baz()
+baz <- function(x) rlang::abort("oh no")
+
+foo()
+#> Error: oh no
+
+rlang::last_trace()
+#> <error/rlang_error>
+#> oh no
+#> Backtrace:
+#>      █
+#>   1. └─global::foo()
+#>   2.   └─global::bar()
+#>   3.     └─1 %>% identity() %>% baz()
+#>   4.       ├─base::withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
+#>   5.       └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
+#>   6.         └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
+#>   7.           └─`_fseq`(`_lhs`)
+#>   8.             └─magrittr::freduce(value, `_function_list`)
+#>   9.               ├─base::withVisible(function_list[[k]](value))
+#>  10.               └─function_list[[k]](value)
+#>  11.                 └─global::baz(.)
+```
+
+This clutter is now completely fixed:
+
+```{r, eval = FALSE}
+foo()
+#> Error: oh no
+
+rlang::last_trace()
+#> <error/rlang_error>
+#> oh no
+#> Backtrace:
+#>     █
+#>  1. ├─global::foo()
+#>  2. │ └─global::bar()
+#>  3. │   └─1 %>% identity() %>% baz()
+#>  4. └─global::baz(.)
+```
+
+Note however that one consequence of having a lazy pipe is that the whole pipeline is pushed on the stack before errors have a chance to be thrown.
+
+```{r, eval = FALSE}
+f1 <- function(x) x
+f2 <- function(x) x
+f3 <- function(x) x
+f4 <- function(x) x
+
+stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
+#> Error in f1(.) : oh no
+
+rlang::last_trace()
+#> <error/rlang_error>
+#> oh no
+#> Backtrace:
+#>     █
+#>  1. ├─stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
+#>  2. ├─global::f4(.)
+#>  3. ├─global::f3(.)
+#>  4. ├─global::f2(.)
+#>  5. └─global::f1(.)
+```
+
+The last function of the pipeline is `f4()`, so that's the first one to be run. It evaluates its argument which is provided by `f3()`, so that's the second function pushed on the stack. And so on until `f1()` needs the result of `stop("oh no")` which causes an error.
+
+
+## Speed
+
+The pipe is now written in C. This greatly improves the performance. Here is a benchmark for the old R implementation:
+
+```{r, eval = FALSE}
+bench::mark(
+  `0` = f1(NULL),
```

Comment:
I'd suggest removing this benchmark, since it's orthogonal to the comparison of interest here (i.e. new vs old, not no pipe vs pipe).